dash-pipeline/dpapp/dash/node.c

```diff
@@ -0,0 +1,354 @@
+/*
+ * Copyright (c) 2015 Cisco and/or its affiliates.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <arpa/inet.h>
+
+#include <vlib/vlib.h>
+#include <vnet/vnet.h>
+#include <vnet/pg/pg.h>
+#include <vnet/ethernet/ethernet.h>
+#include <vppinfra/error.h>
+#include <dash/dash.h>
+
+typedef struct
+{
+  u32 next_index;
+  u32 sw_if_index;
+  u8 new_src_mac[6];
+  u8 new_dst_mac[6];
+} dash_trace_t;
+
+
+/* packet trace format function */
+static u8 *
+format_dash_trace (u8 * s, va_list * args)
+{
+  CLIB_UNUSED (vlib_main_t * vm) = va_arg (*args, vlib_main_t *);
+  CLIB_UNUSED (vlib_node_t * node) = va_arg (*args, vlib_node_t *);
+  dash_trace_t *t = va_arg (*args, dash_trace_t *);
+
+  s = format (s, "DASH: sw_if_index %d, next index %d\n",
+	      t->sw_if_index, t->next_index);
+  s = format (s, "  new src %U -> new dst %U",
+	      format_mac_address, t->new_src_mac,
+	      format_mac_address, t->new_dst_mac);
+
+  return s;
+}
+
+extern vlib_node_registration_t dash_node;
+
+#define foreach_dash_error \
+_(SWAPPED, "Mac swap packets processed")
+
+typedef enum
+{
+#define _(sym,str) DASH_ERROR_##sym,
+  foreach_dash_error
+#undef _
+    DASH_N_ERROR,
+} dash_error_t;
+
+static char *dash_error_strings[] = {
+#define _(sym,string) string,
+  foreach_dash_error
+#undef _
+};
+
+typedef enum
+{
+  DASH_NEXT_INTERFACE_OUTPUT,
+  DASH_N_NEXT,
+} dash_next_t;
+
+#define foreach_mac_address_offset              \
+_(0)                                            \
+_(1)                                            \
+_(2)                                            \
+_(3)                                            \
+_(4)                                            \
+_(5)
+
+VLIB_NODE_FN (dash_node) (vlib_main_t * vm, vlib_node_runtime_t * node,
+			    vlib_frame_t * frame)
+{
+  u32 n_left_from, *from, *to_next;
+  dash_next_t next_index;
+  u32 pkts_swapped = 0;
+
+  from = vlib_frame_vector_args (frame);
+  n_left_from = frame->n_vectors;
+  next_index = node->cached_next_index;
+
+  while (n_left_from > 0)
+    {
+      u32 n_left_to_next;
+
+      vlib_get_next_frame (vm, node, next_index, to_next, n_left_to_next);
+
+      while (n_left_from >= 4 && n_left_to_next >= 2)
+	{
+	  u32 next0 = DASH_NEXT_INTERFACE_OUTPUT;
+	  u32 next1 = DASH_NEXT_INTERFACE_OUTPUT;
+	  u32 sw_if_index0, sw_if_index1;
+	  u8 tmp0[6], tmp1[6];
```

Comment:
if multiple packets are processed in 1 iterator, at least we should create a struct context for each packet instead of doing *0,1. if we really like to do it in the vector way, we should also leverage array, so SIMD instructions can be generated.

Reply:
also do we really need to process multiple packets at once? although the framework allow us to do, but I don't think this performance will impact us that much.

Reply:
For simplicity, removed multiple packets processing in 1 iterator.